src/lib/core/fiber.c

```diff
@@ -894,13 +888,97 @@ fiber_recycle(struct fiber *fiber)
 	unregister_fid(fiber);
 	fiber->fid = 0;
 	region_free(&fiber->gc);
+#ifdef ENABLE_BACKTRACE
+	fiber->parent_bt = NULL;
+	fiber->leak_bt = NULL;
+	region_set_callbacks(&fiber->gc, NULL, NULL, NULL);
+#endif /* ENABLE_BACKTRACE */
+	fiber->gc_internal = 0;
+	fiber->debug_enabled = false;
 	if (!has_custom_stack) {
 		rlist_move_entry(&cord()->dead, fiber, link);
 	} else {
 		cord_add_garbage(cord(), fiber);
 	}
 }
 
+#ifdef ENABLE_BACKTRACE
+/**
+ * Called on allocation on region gc. Saves allocation caller backtrace
+ * to report it later if region gc leak is found.
+ */
+static void
+fiber_on_gc_alloc(const struct region *region,
+		  size_t size,
+		  void *opaque)
+{
+	(void)region;
+	(void)size;
+	struct fiber *fiber = opaque;
+
+	assert(region_used(&fiber->gc) >= fiber->gc_internal);
+	if (region_used(&fiber->gc) == fiber->gc_internal) {
+		/* 1 is to skip not interesting frames. */
+		backtrace_collect(fiber->leak_bt, NULL, 1);
+		assert(fiber->leak_bt->frame_count > 0);
+	}
+}
+
+/**
+ * Called on region gc truncation. Resets previously saved backtrace so
+ * it won't be falsely reported in case of programmatic mistake.
+ */
+static void
+fiber_on_gc_truncate(const struct region *region,
+		     size_t used,
+		     void *opaque)
+{
+	(void)region;
+	struct fiber *fiber = opaque;
+
+	assert(used >= fiber->gc_internal);
+	if (used == fiber->gc_internal)
+		fiber->leak_bt->frame_count = 0;
+}
+#endif /* ENABLE_BACKTRACE */
+
+void
+fiber_check_gc(void)
+{
+	struct fiber *fiber = fiber();
+
+	assert(region_used(&fiber->gc) >= fiber->gc_internal);
+	if (region_used(&fiber->gc) == fiber->gc_internal)
+		return;
+
+#ifdef ENABLE_BACKTRACE
+	if (fiber->debug_enabled) {
```

Comment:
Do you really need this flag? Can't you check if `leak_bt` is NULL instead?

Reply:
Yep, dropped.